File: attentive-android-sdk/src/main/java/com/attentive/androidsdk/events/ProductViewEvent.java
package com.attentive.androidsdk.events;

import com.attentive.androidsdk.ParameterValidation;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
import java.util.ArrayList;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@JsonDeserialize(builder = ProductViewEvent.Builder.class)
public final class ProductViewEvent extends Event {
    @NotNull
    private final List<Item> items;
    @Nullable
    private final String deeplink;

    private ProductViewEvent(@NotNull List<Item> items) {
        this.items = items;
        this.deeplink = null;
    }

    private ProductViewEvent(@NotNull Builder builder) {
        this.items = builder.items;
        this.deeplink = builder.deeplink;
    }

    @JsonPOJOBuilder(withPrefix = "")
    public static final class Builder {
        @NotNull
        private List<Item> items = new ArrayList<>();
        @Nullable
        private String deeplink;

        /**
         * @deprecated As of release 1.0.0-beta01, replaced by standard builder methods
         */
        @Deprecated(since = "1.0.0-beta01", forRemoval = true)
        @JsonCreator
        public Builder(@JsonProperty("items") List<Item> items) {
            ParameterValidation.verifyNotEmpty(items, "items");

            this.items = new ArrayList<>(items);
            this.deeplink = null;
        }

        public Builder() {}

        public Builder items(@NotNull List<Item> items) {
            this.items = items;
            return this;
        }

        public Builder deeplink(@Nullable String deeplink) {
            this.deeplink = deeplink;
            return this;
        }

        public ProductViewEvent buildIt() {
            return new ProductViewEvent(this);
        }

        /**
         * @deprecated As of release 1.0.0-beta01, replaced by {@link ProductViewEvent.Builder#buildIt}
         * Only to be used if using the new builder approach not the deprecated one.
         */
        @Deprecated(since = "1.0.0-beta01", forRemoval = true)
        public ProductViewEvent build() {
            return new ProductViewEvent(this.items);
        }
    }

    @NotNull
    public List<Item> getItems() {
        return items;
    }

    @Nullable
    public String getDeeplink() {
        return deeplink;
    }
}

File: attentive-android-sdk/src/main/java/com/attentive/androidsdk/events/ProductViewEvent.kt
package com.attentive.androidsdk.events

import com.attentive.androidsdk.ParameterValidation
import kotlinx.serialization.Serializable

/**
 * A product view event. Fires when the user views one or more products (e.g. lands on a PDP).
 *
 * @property items The items viewed. Must be non-empty.
 * @property deeplink Optional deeplink associated with the event. Propagated to the backend
 *   so campaigns can re-target the user with the same URL.
 */
@Serializable
data class ProductViewEvent(
    val items: List<Item>,
    val deeplink: String? = null,
) : Event() {
    init {
        ParameterValidation.verifyNotEmpty(items, "items")
    }

    companion object {
        @Deprecated(
            "As of release 1.0.0-beta01, replaced by standard builder methods",
            ReplaceWith("ProductViewEvent.Builder().items(items).build()"),
        )
        fun create(items: List<Item>): ProductViewEvent {
            ParameterValidation.verifyNotEmpty(items, "items")
            return ProductViewEvent(items)
        }
    }

    @Serializable
    class Builder {
        var items: List<Item> = emptyList()
            private set
        var deeplink: String? = null
            private set

        fun items(items: List<Item>): Builder {
            this.items = items
            return this
        }

        fun deeplink(deeplink: String?): Builder {
            this.deeplink = deeplink
            return this
        }

        /**
         * @throws IllegalArgumentException if [items] is empty.
         */
        fun build(): ProductViewEvent {
            return ProductViewEvent(items, deeplink)
        }
    }
}
